src/flag_attn/split_kv.py

```diff
@@ -0,0 +1,325 @@
+        tl.store(o_ptrs, acc.to(input_dtype), mask=mask_m[:, None], cache_modifier=".cg")
+
+@triton.jit
+def _fwd_combine_kv_splits(
+    multiple_o, multiple_l,
+    final_o, final_l,
+    stride_mul_oz, stride_mul_oh, stride_mul_os, stride_mul_om, stride_mul_ok,
+    stride_fin_oz, stride_fin_oh, stride_fin_om, stride_fin_ok,
+    Z, H, M, S,
+    BLOCK_M: tl.constexpr, BLOCK_DMODEL: tl.constexpr,
+    DIVISIBLE_M: tl.constexpr,
+):
+    start_m = tl.program_id(0)
+    offs_h = tl.program_id(1)
+    offs_z = tl.program_id(2)
+
+    # offset
+    multiple_o += offs_z * stride_mul_oz + offs_h * stride_mul_oh # (B, H, S, M, D)
+    multiple_l += (offs_z * H + offs_h) * S * M # (B, H, S, M)
+
+    offs_m = start_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    if not DIVISIBLE_M:
+        mask_m = offs_m < M
+
+    # 1st loop: online logsumexp to save a swipe
+    m = tl.full([BLOCK_M], value=float("-inf"), dtype=tl.float32)
+    acc = tl.full([BLOCK_M], value=float(0.0), dtype=tl.float32)
+    l_ptrs = multiple_l + offs_m
+    for _ in range(0, S):
+        if DIVISIBLE_M:
+            l = tl.load(l_ptrs)
+        else:
+            l = tl.load(l_ptrs, mask=mask_m)
+        m_new = tl.maximum(m, l)
+        acc = acc * tl.exp(m - m_new) + tl.exp(l - m_new)
+        m = m_new
+        l_ptrs += M
+    l_acc = m + tl.log(acc)
+
+    # 2rd loop to rescale and accumulate o
+    o_acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
+    l_ptrs = multiple_l + offs_m
+    offs_k = tl.arange(0, BLOCK_DMODEL)
+    o_ptrs = multiple_o + offs_m[:, None] * stride_mul_om + offs_k[None, :] * stride_mul_ok
+    for _ in range(0, S):
+        l = tl.load(l_ptrs, mask=offs_m < M)
+        rescale = tl.exp(l - l_acc)
+        if DIVISIBLE_M:
+            o = tl.load(o_ptrs, )
+        else:
+            o = tl.load(o_ptrs, mask=mask_m[:, None])
+        o_acc += o * rescale[:, None]
+
+        l_ptrs += M
+        o_ptrs += stride_mul_os
+
+    # write back
+    final_o += offs_z * stride_fin_oz + offs_h * stride_fin_oh
+    final_l += (offs_z * H + offs_h) * M
+    a_ptrs = final_o + offs_m[:, None] * stride_fin_om + offs_k * stride_fin_ok
+    b_ptrs = final_l + offs_m
+
+    if DIVISIBLE_M:
+        tl.store(a_ptrs, o_acc)
+        tl.store(b_ptrs, l_acc)
+    else:
+        tl.store(a_ptrs, o_acc, mask=mask_m[:, None])
+        tl.store(b_ptrs, l_acc, mask=mask_m)
+
+def get_fwd_config(B, H, M, N, D, causal):
+    # BLOCK_M, BLOCK_N, num_stages, num_warps
+    return (16, 128, 1, 4)
+
+# this function is adapted from https://github.com/Dao-AILab/flash-attention/blob/61a777247900f6c2a37376f3ffd7134385fdc95c/csrc/flash_attn/flash_api.cpp#L235
+def num_splits_herustic(B, H, M, N, BLOCK_M, BLOCK_N, num_sms, max_splits):
+    num_blocks_without_split_kv = B * H * triton.cdiv(M, BLOCK_M)
+    if num_blocks_without_split_kv >= 0.8 * num_sms:
+        return 1
+
+    num_n_blocks = triton.cdiv(N, BLOCK_N)
+    def num_split_avaiable(s):
```

Comment:
available?

Reply:
When spliting `num_n_blocks` into `s` splits, there may be some splits(the last split, for example) without valid workload. 

For example, when spliting `64` blocks into `12` splits, with each split processing `cdiv(64, 12) = 6` blocks, the last split is  empty. So spliting `64` blocks into `12` splits is not available in this sense.